scripts/deployment/phase1-common/2_deploy_implementations.ts

```diff
@@ -113,98 +168,242 @@ async function main() {
+    brokerImplAddr = await upgrades.deployImplementation(BrokerImplFactory, {
+      kind: 'uups',
+    })
+  } else {
+    brokerImplAddr = await upgrades.prepareUpgrade(
+      prevDeployments.implementations.components.broker,
+      BrokerImplFactory,
+      {
+        kind: 'uups',
+      }
+    )
+  }
 
-  writeComponentDeployment(deployments, deploymentFilename, 'broker', brokerImpl.address, 'Broker')
+  brokerImpl = <BrokerP1>await ethers.getContractAt('BrokerP1', brokerImplAddr)
+
+  writeComponentDeployment(
+    deployments,
+    deploymentFilename,
+    'broker',
+    brokerImpl.address,
+    'Broker',
+    prevDeployments.implementations.components.broker
+  )
 
   // 5. *********** Distributor *************/
   const DistribImplFactory = await ethers.getContractFactory('DistributorP1')
-  distribImpl = <DistributorP1>await DistribImplFactory.connect(burner).deploy()
-  await distribImpl.deployed()
+  let distribImplAddr = ''
+  if (!upgrade) {
+    distribImplAddr = await upgrades.deployImplementation(DistribImplFactory, {
+      kind: 'uups',
+    })
+  } else {
+    distribImplAddr = await upgrades.prepareUpgrade(
+      prevDeployments.implementations.components.distributor,
+      DistribImplFactory,
+      {
+        kind: 'uups',
+      }
+    )
+  }
+
+  distribImpl = <DistributorP1>await ethers.getContractAt('DistributorP1', distribImplAddr)
 
   writeComponentDeployment(
     deployments,
     deploymentFilename,
     'distributor',
     distribImpl.address,
-    'Distributor'
+    'Distributor',
+    prevDeployments.implementations.components.distributor
   )
 
   // 6. *********** Furnace *************/
   const FurnaceImplFactory = await ethers.getContractFactory('FurnaceP1')
-  furnaceImpl = <FurnaceP1>await FurnaceImplFactory.connect(burner).deploy()
-  await furnaceImpl.deployed()
+  let furnaceImplAddr = ''
+  if (!upgrade) {
+    furnaceImplAddr = await upgrades.deployImplementation(FurnaceImplFactory, {
+      kind: 'uups',
+    })
+  } else {
+    furnaceImplAddr = await upgrades.prepareUpgrade(
+      prevDeployments.implementations.components.furnace,
+      FurnaceImplFactory,
+      {
+        kind: 'uups',
+      }
+    )
+  }
+
+  furnaceImpl = <FurnaceP1>await ethers.getContractAt('FurnaceP1', furnaceImplAddr)
 
   writeComponentDeployment(
     deployments,
     deploymentFilename,
     'furnace',
     furnaceImpl.address,
-    'Furnace'
+    'Furnace',
+    prevDeployments.implementations.components.furnace
   )
 
   // 7. *********** RevenueTrader *************/
 
   const RevTraderImplFactory = await ethers.getContractFactory('RevenueTraderP1')
-  revTraderImpl = <RevenueTraderP1>await RevTraderImplFactory.connect(burner).deploy()
-  await revTraderImpl.deployed()
+  let revTraderImplAddr = ''
+  if (!upgrade) {
+    revTraderImplAddr = await upgrades.deployImplementation(RevTraderImplFactory, {
+      kind: 'uups',
+    })
+  } else {
+    // We can use any of the two addresses (should be equal)
```

Comment:
*should*, but might not be. probably worth at least verifying that they are equal? or making sure we verify against both?